src/sparse_int_set.jl

```diff
@@ -0,0 +1,291 @@
+import Base: @propagate_inbounds, zip
+
+const INT_PER_PAGE = div(ccall(:jl_getpagesize, Clong, ()), sizeof(Int))
+
+
+#TODO: Batch creation and allocation
+mutable struct SparseIntSet
+    packed ::Vector{Int}
+    reverse::Vector{Vector{Int}}
+    counters::Vector{Int}  # counts the number of real elements in each page of reverse. 
+end
+
+SparseIntSet() = SparseIntSet(Int[], Vector{Int}[], Int[])
+
+SparseIntSet(indices) = union!(SparseIntSet(), indices)
+
+eltype(::Type{SparseIntSet}) = Int
+
+empty(::SparseIntSet) = SparseIntSet()
+
+function empty!(s::SparseIntSet)
+    empty!(s.packed)
+    empty!(s.reverse)
+    empty!(s.counters)
+    return s
+end
+
+isempty(s::SparseIntSet) = isempty(s.packed)
+
+copy(s::SparseIntSet) = copy!(SparseIntSet(), s)
+
+function copy!(to::SparseIntSet, from::SparseIntSet)
+    to.packed   = copy(from.packed)
+    to.reverse  = deepcopy(from.reverse)
+    to.counters = copy(from.counters)
+    return to
+end
+
+function pageid_offset(s::SparseIntSet, i)
+    pageid = div(i - 1, INT_PER_PAGE) + 1
+    return pageid, (i - 1) & (INT_PER_PAGE - 1) + 1
+end
+
+function in(i, s::SparseIntSet)
+    pageid, offset = pageid_offset(s, i)
+    isassigned(s.reverse, pageid) && @inbounds s.reverse[pageid][offset] != 0
+end
+
+length(s::SparseIntSet) = length(s.packed)
+
+# This makes sure that when adding (pushing) an Int,
+# it's respective page is allocated and put at the index of the reverse such that
+# pageid_offset works as intended.
+# Pages will be allocated only once, when pushing an Int that belongs to them.
+# Other not used pages (created during resize!) will be undefs until one Int belonging to them gets added. 
+function assure!(s::SparseIntSet, pageid)
+    if pageid > length(s.reverse)
+        diff = pageid - length(s.reverse)
+
+        resize!(s.reverse, pageid - 1)
```

Comment:
Probably better not to do that?  `push!` will resize for you.
And it has heuristics to better handle the resizing

Reply:
Not sure I understand what you mean, do you mean to move `assure!` into `push!` I guess that's ok since I don't think it's used anywhere else

Reply:
no, I mean why are we doing `resize!` then `push!` ?

Reply:
Before this was to generate the `undef`s but now indeed i'm just using `push!` after `sizehint!`, I don't know if that's optimal

Reply:
it is

Reply:
did a quick benchmark with doing `resize!` followed by assignment to index, difference is basically swallowed by noise on the benchmark it seems.